### PR TITLE
Decouple Script from Storage

### PR DESCRIPTION
**Pull request by MikeStall**
Decouple Script runtime from storage. to enable the OnPremise scenario. 
Make BlobLeaseManager relying on SDK's new IDistributedLockManager, rather than directly rely on storage. 
This ports Script's existing BlobLeaseManagerTests and keeps them running, leveraging the underlying Blob implementation. 

Resolves https://github.com/Azure/azure-webjobs-sdk-script/issues/1584 

This needs SDK changes from https://github.com/Azure/azure-webjobs-sdk/pull/1184 

Update nugget references. 
Include test fix. Longer fix needed: https://github.com/Azure/azure-webjobs-sdk-script/issues/1674

### CODE REVIEW

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 1]
```diff
@@ -428,9 +424,16 @@ protected virtual void Initialize()
 
                 LoadCustomExtensions();
 
+                // Do this after we've loaded the custom extensions. That gives an extension an opportunity to plug in their own implementations.
+                if (storageString != null)
+                {
+                    var lockManager = this.ScriptConfig.HostConfig.GetService<IDistributedLockManager>();
```
Comment (fabiocav) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Really wish we were not committing further to the pattern of pulling stuff from the service provider directly like this. We'll need to rework how DI is done, so if giving extensions the ability to inject an `IDistributedLockManager` is just a nice to have, let's avoid this here so we can have it properly done when we get to it.

Also, is the core SDK registering a default implementation of a `IDistributedLockManager` implementation before this point here? Replacing services in different stages in this process makes it easy to introduce bugs like what we had with configuration and logging. If we want extensions to participate in DI, we should have that properly exposed and allow them to do so in the right place.

File: src/WebJobs.Script/Host/BlobLeaseManager.cs [hunk 2]
```diff
@@ -202,48 +191,9 @@ private void ProcessLeaseError(string reason)
...
-                }
-            }
-
-            return blob;
-        }
-
         private void ResetLease()
         {
-            LeaseId = null;
+            this.LockHandle = null;
```
Comment (fabiocav) [inline src/WebJobs.Script/Host/BlobLeaseManager.cs]: nit: remove `this.` throughout

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 3]
```diff
@@ -428,9 +424,16 @@ protected virtual void Initialize()
...
+                    blobManagerCreation = BlobLeaseManager.Create(lockManager, TimeSpan.FromSeconds(15), hostConfig.HostId, InstanceId, TraceWriter, hostConfig.LoggerFactory);
+                }
+
                 // Create the lease manager that will keep handle the primary host blob lease acquisition and renewal
                 // and subscribe for change notifications.
-                _blobLeaseManager = blobManagerCreation.GetAwaiter().GetResult();
+                _blobLeaseManager = blobManagerCreation;
```
Comment (fabiocav) [inline src/WebJobs.Script/Host/ScriptHost.cs]: nit: With this implementation, we no longer have the need to assign it to a temporary variable here as we're not kicking the creation process in the background and getting the result later, so we can get rid of it.

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 4]
```diff
@@ -500,7 +503,8 @@ private void LoadCustomExtensions()
             // There's a single script binding instance that services all extensions.
             // give that script binding the metadata for all loaded extensions so it can dispatch to them.
             var generalProvider = ScriptConfig.BindingProviders.OfType<GeneralScriptBindingProvider>().First();
-            generalProvider.CompleteInitialization();
+            var metadataProvider = this.CreateMetadataProvider();
```
Comment (fabiocav) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Is this an unrelated change?

File: test/WebJobs.Script.Tests.Integration/BlobLeaseManagerTests.cs [hunk 5]
```diff
@@ -65,12 +72,14 @@ public class BlobLeaseManagerTests
 
             resetEvent.Wait(TimeSpan.FromSeconds(15));
             bool hasLease = manager.HasLease;
-            string actualLeaseId = manager.LeaseId;
+
+            // string actualLeaseId = manager.LeaseId;
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests.Integration/BlobLeaseManagerTests.cs]: nit: remove commented out code.

File: test/WebJobs.Script.Tests.Integration/BlobLeaseManagerTests.cs [hunk 6]
```diff
@@ -65,12 +72,14 @@ public class BlobLeaseManagerTests
...
             manager.Dispose();
 
             Assert.True(resetEvent.IsSet);
             Assert.True(hasLease, $"{nameof(BlobLeaseManager.HasLease)} was not correctly set to 'true' when lease was acquired.");
-            Assert.Equal(instanceId, actualLeaseId);
+
+            // Assert.Equal(instanceId, actualLeaseId);
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests.Integration/BlobLeaseManagerTests.cs]: nit: remove commented out code.

File: src/WebJobs.Script/Host/BlobLeaseManager.cs [hunk 7]
```diff
@@ -123,62 +126,48 @@ private void ProcessLeaseTimerTick(object state)
...
+                _lastRenewalLatency = _lastRenewal - requestStart;
+
+                string message = $"Host lock lease acquired by instance ID '{_instanceId}'.";
+                _traceWriter.Info(message);
+                _logger?.LogInformation(message);
+
+                // We've successfully acquired the lease, change the timer to use our renewal interval
+                SetTimerInterval(_renewalInterval);
             }
         }
```
Comment (fabiocav) [inline src/WebJobs.Script/Host/BlobLeaseManager.cs]: `GetBlobName` below should be removed.

File: test/WebJobs.Script.Tests.Integration/BlobLeaseManagerTests.cs [hunk 8]
```diff
@@ -119,76 +130,6 @@ public class BlobLeaseManagerTests
         }
 
         [Fact]
-        public void AcquiringLease_WithServerError_LogsAndRetries()
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests.Integration/BlobLeaseManagerTests.cs]: Are we logging anything in the failure path? I believe we're losing this information. Can we have a test to validate that?

File: test/WebJobs.Script.Tests/ScriptHostTests.cs [hunk 9]
```diff
@@ -451,7 +451,7 @@ public void ApplyConfiguration_Queues()
             scriptConfig.HostConfig.HostConfigMetadata = config;
             TraceWriter traceWriter = new TestTraceWriter(TraceLevel.Verbose);
 
-            scriptConfig.HostConfig.CreateMetadataProvider(); // will cause extensions to initialize and consume config metadata.
+            new JobHost(scriptConfig.HostConfig).CreateMetadataProvider(); // will cause extensions to initialize and consume config metadata.
```
Comment (fabiocav) [inline test/WebJobs.Script.Tests/ScriptHostTests.cs]: There are two sets of changes in the PR, not much context on the second one. Can you provide more information o this?

File: src/WebJobs.Script/Binding/GeneralScriptBindingProvider.cs [hunk 10]
```diff
@@ -29,9 +29,9 @@ internal class GeneralScriptBindingProvider : ScriptBindingProvider
         // The constructor is fixed and ScriptBindingProvider are instantated for us by the Script runtime.
         // Extensions may get registered after this class is instantiated.
         // So we need a final call that lets us get the tooling snapshot of the graph after all extensions are set.
-        public void CompleteInitialization()
+        public void CompleteInitialization(IJobHostMetadataProvider metadataProvider)
         {
-            this._metadataProvider = this.Config.CreateMetadataProvider();
```
Comment (MikeStall) [inline src/WebJobs.Script/Binding/GeneralScriptBindingProvider.cs]: Yes, although it could have been factored out to a separate change. We had to do some juggling to allow Script to get access to the new IDistributedLockMAnager,   (see [https://github.com/Azure/azure-webjobs-sdk-script/pull/1615](https://github.com/Azure/azure-webjobs-sdk-script/pull/1615) ) 

---
In reply to: [123669490](https://github.com/Azure/azure-webjobs-sdk-script/pull/1615#discussion_r123669490) [](ancestors = 123669490)

File: src/WebJobs.Script/Host/BlobLeaseManager.cs [hunk 11]
```diff
@@ -123,62 +126,48 @@ private void ProcessLeaseTimerTick(object state)
...
+                    ProcessLeaseError($"Another host has acquired the lease. The last successful renewal completed at {lastRenewalFormatted} ({millisecondsSinceLastSuccess} milliseconds ago) with a duration of {lastRenewalMilliseconds} milliseconds.");
                 }
             }
-            catch (StorageException exc)
+            else
             {
-                if (exc.RequestInformation.HttpStatusCode == 409)
+                // LockManager will handle various cases like 500s, 404s, retry if container doesn't exist, etc.
+                string proposedLeaseId = _instanceId;
+                this.LockHandle = await _lockManager.TryLockAsync(null, lockName, _instanceId, proposedLeaseId, _leaseTimeout, CancellationToken.None);
```
Comment (MikeStall) [inline src/WebJobs.Script/Host/BlobLeaseManager.cs]: yes :( Not needed in the Script or Listener case; needed to support Singleton.Account. 

---
In reply to: [123670601](https://github.com/Azure/azure-webjobs-sdk-script/pull/1615#discussion_r123670601) [](ancestors = 123670601)

Comment (MikeStall) [inline src/WebJobs.Script/Host/BlobLeaseManager.cs]: For diagnostics, we stamp a lock with an "owner".  SDK uses this aggressively so that you know which function instance holds the lock.  Could be null in this case. 

---
In reply to: [123670650](https://github.com/Azure/azure-webjobs-sdk-script/pull/1615#discussion_r123670650) [](ancestors = 123670650)

File: src/WebJobs.Script/Host/BlobLeaseManager.cs [hunk 12]
```diff
@@ -123,62 +126,48 @@ private void ProcessLeaseTimerTick(object state)
...
+                if (this.LockHandle == null)
```
Comment (MikeStall) [inline src/WebJobs.Script/Host/BlobLeaseManager.cs]: What's the duplication?  I thought it was cleaner to directly refer to the return value here. 

---
In reply to: [123670281](https://github.com/Azure/azure-webjobs-sdk-script/pull/1615#discussion_r123670281) [](ancestors = 123670281)

File: src/WebJobs.Script/Host/BlobLeaseManager.cs (hunk 7 above)
Comment (MikeStall) [inline src/WebJobs.Script/Host/BlobLeaseManager.cs]: We still need it, but perhaps we could give it a better name. This is the "lockId" - ie, the name of what's being locked. 

---
In reply to: [123674089](https://github.com/Azure/azure-webjobs-sdk-script/pull/1615#discussion_r123674089) [](ancestors = 123674089)

File: src/WebJobs.Script/Host/BlobLeaseManager.cs (hunk 2 above)
Comment (MikeStall) [inline src/WebJobs.Script/Host/BlobLeaseManager.cs]: Can we add a Style cop error for that?

---
In reply to: [123673112](https://github.com/Azure/azure-webjobs-sdk-script/pull/1615#discussion_r123673112) [](ancestors = 123673112)

File: src/WebJobs.Script/Host/BlobLeaseManager.cs [hunk 13]
```diff
@@ -258,7 +208,8 @@ private void TryReleaseLeaseIfOwned()
             {
                 if (HasLease)
                 {
-                    _lockBlob.ReleaseLease(new AccessCondition { LeaseId = LeaseId });
+                    Task.Run(() => _lockManager.ReleaseLockAsync(_lockHandle, CancellationToken.None)).GetAwaiter().GetResult();
```
Comment (MikeStall) [inline src/WebJobs.Script/Host/BlobLeaseManager.cs]: We can't - this is called from Dispose, which is non-async.  :(    

---
In reply to: [123671702](https://github.com/Azure/azure-webjobs-sdk-script/pull/1615#discussion_r123671702) [](ancestors = 123671702)

File: test/WebJobs.Script.Tests/ScriptHostTests.cs (hunk 9 above)
Comment (MikeStall) [inline test/WebJobs.Script.Tests/ScriptHostTests.cs]: [https://github.com/Azure/azure-webjobs-sdk/issues/1201](https://github.com/Azure/azure-webjobs-sdk/issues/1201)

---
In reply to: [123674427](https://github.com/Azure/azure-webjobs-sdk-script/pull/1615#discussion_r123674427) [](ancestors = 123674427)

File: test/WebJobs.Script.Tests.Integration/BlobLeaseManagerTests.cs (hunk 8 above)
Comment (MikeStall) [inline test/WebJobs.Script.Tests.Integration/BlobLeaseManagerTests.cs]: These tests are testing cases that are now handled by the SDK.  

---
In reply to: [123674215](https://github.com/Azure/azure-webjobs-sdk-script/pull/1615#discussion_r123674215) [](ancestors = 123674215)

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 1 above)
Comment (MikeStall) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Yes, this is committing further to allowing Script to pull the same services that SDK is using. 
SDK provides a default IDistributedLockManager.  In the on-prem case, an extension will override and inject their own. 

---
In reply to: [123672594](https://github.com/Azure/azure-webjobs-sdk-script/pull/1615#discussion_r123672594) [](ancestors = 123672594)

File: src/WebJobs.Script/Host/BlobLeaseManager.cs [hunk 14]
```diff
@@ -26,22 +26,25 @@ internal sealed class BlobLeaseManager : IDisposable
...
+
         private bool _disposed;
         private bool _processingLease;
         private DateTime _lastRenewal;
         private TimeSpan _lastRenewalLatency;
 
-        internal BlobLeaseManager(ICloudBlob lockBlob, TimeSpan leaseTimeout, string hostId, string instanceId, TraceWriter traceWriter,
+        private IDistributedLockManager _lockManager;
+
+        internal BlobLeaseManager(IDistributedLockManager lockManager, TimeSpan leaseTimeout, string hostId, string instanceId, TraceWriter traceWriter,
```
Comment (MikeStall) [inline src/WebJobs.Script/Host/BlobLeaseManager.cs]: Will rename to "PrimaryHostCoordinator" 

---
In reply to: [123669617](https://github.com/Azure/azure-webjobs-sdk-script/pull/1615#discussion_r123669617) [](ancestors = 123669617)

File: src/WebJobs.Script/Host/BlobLeaseManager.cs (hunk 11 above)
Comment (MikeStall) [inline src/WebJobs.Script/Host/BlobLeaseManager.cs]: we don't have one. 

---
In reply to: [123670562](https://github.com/Azure/azure-webjobs-sdk-script/pull/1615#discussion_r123670562) [](ancestors = 123670562)

File: src/WebJobs.Script/Host/BlobLeaseManager.cs [hunk 15]
```diff
@@ -123,62 +126,48 @@ private void ProcessLeaseTimerTick(object state)
...
+                try
                 {
-                    await _lockBlob.RenewLeaseAsync(new AccessCondition { LeaseId = LeaseId });
+                    await _lockManager.RenewAsync(this.LockHandle, CancellationToken.None);
+
                     _lastRenewal = DateTime.UtcNow;
                     _lastRenewalLatency = _lastRenewal - requestStart;
                 }
-                else
+                catch
```
Comment (MikeStall) [inline src/WebJobs.Script/Host/BlobLeaseManager.cs]: No. I chose against introducing a new exception type (LockRenewalException) because I don't think it actually solves anything here. We'd still need to handle the general case. And we'd handle general and specific exception the same way. And it places an additional burden on implementors. 

---
In reply to: [123671286](https://github.com/Azure/azure-webjobs-sdk-script/pull/1615#discussion_r123671286) [](ancestors = 123671286)

File: src/WebJobs.Script/Host/BlobLeaseManager.cs [hunk 16]
```diff
@@ -78,24 +81,24 @@ private set
...
+            TimeSpan leaseTimeout,
+            string hostId,
+            string instanceId,
+            TraceWriter traceWriter,
+            ILoggerFactory loggerFactory,
+            TimeSpan? renewalInterval = null)
         {
             if (leaseTimeout.TotalSeconds < 15 || leaseTimeout.TotalSeconds > 60)
             {
                 throw new ArgumentOutOfRangeException(nameof(leaseTimeout), $"The {nameof(leaseTimeout)} should be between 15 and 60 seconds");
```
Comment (MikeStall) [inline src/WebJobs.Script/Host/BlobLeaseManager.cs]: Good point. Although there's only one callsite here (in ScriptHost.cs) and the value passed in is a compile-time constant. So this check is of questionable value. 

---
In reply to: [123670795](https://github.com/Azure/azure-webjobs-sdk-script/pull/1615#discussion_r123670795) [](ancestors = 123670795)

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 17]
```diff
@@ -52,7 +52,7 @@ public class ScriptHost : JobHost
         private Action _shutdown;
         private AutoRecoveringFileSystemWatcher _debugModeFileWatcher;
         private ImmutableArray<string> _directorySnapshot;
-        private BlobLeaseManager _blobLeaseManager;
+        private PrimaryHostCoordinator _blobLeaseManager;
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: nit: update to make fields match type
